File: lib/models/entry.dart
import 'package:uuid/uuid.dart';

const _uuid = Uuid();

class Entry {
  final String id;
  String title;
  String category;
  String username;
  String password;
  String url;
  String notes;
  bool isFavorite;
  final DateTime createdAt;
  DateTime updatedAt;

  Entry({
    String? id,
    required this.title,
    required this.category,
    required this.username,
    required this.password,
    this.url = '',
    this.notes = '',
    this.isFavorite = false,
    DateTime? createdAt,
    DateTime? updatedAt,
  })  : id = id ?? _uuid.v4(),
        createdAt = createdAt ?? DateTime.now(),
        updatedAt = updatedAt ?? DateTime.now();

  Entry copyWith({
    String? title,
    String? category,
    String? username,
    String? password,
    String? url,
    String? notes,
    bool? isFavorite,
  }) =>
      Entry(
        id: id,
        title: title ?? this.title,
        category: category ?? this.category,
        username: username ?? this.username,
        password: password ?? this.password,
        url: url ?? this.url,
        notes: notes ?? this.notes,
        isFavorite: isFavorite ?? this.isFavorite,
        createdAt: createdAt,
        updatedAt: DateTime.now(),
      );

  factory Entry.fromJson(Map<String, dynamic> json) => Entry(
        id: json['id'] as String,
        title: json['title'] as String,
        category: json['category'] as String? ?? 'Autres',
        username: json['username'] as String? ?? '',
        password: json['password'] as String? ?? '',
        url: json['url'] as String? ?? '',
        notes: json['notes'] as String? ?? '',
        isFavorite: json['isFavorite'] as bool? ?? false,
        createdAt: DateTime.parse(json['createdAt'] as String),
        updatedAt: DateTime.parse(json['updatedAt'] as String),
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'category': category,
        'username': username,
        'password': password,
        'url': url,
        'notes': notes,
        'isFavorite': isFavorite,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };
}

File: app/src/main/java/com/filestech/pass_tech/core/model/Entry.kt
package com.filestech.pass_tech.core.model

enum class EntryType(val wireName: String) {
    PASSWORD("password"),
    NOTE("note"),
    CARD("card"),
    ;

    companion object {
        /** As in the Flutter app, an unknown or missing type reads as a password entry. */
        fun fromWireName(name: String?): EntryType = entries.firstOrNull { it.wireName == name } ?: PASSWORD
    }
}

/**
 * One vault entry. The field set and the JSON names are those of the Flutter app (2.7.1,
 * `lib/models/entry.dart`), so that backups move both ways without loss.
 *
 * `category` is stored as the canonical FRENCH name (`Banque`, `Réseaux sociaux`...) and translated
 * at display time, like in the Flutter app: translating it in storage would break the category of
 * every entry the day the language changes.
 *
 * [otherDomains] is the one field 2.7.1 does not know. It is written only when it holds something,
 * so an entry that never used it is byte-identical to what 2.7.1 wrote, and a 2.7.1 reader — which
 * ignores keys it does not know — opens a 3.0 backup whatever happens.
 */
data class Entry(
    val id: String,
    val type: EntryType = EntryType.PASSWORD,
    val title: String,
    val category: String,
    val username: String = "",
    val password: String = "",
    val url: String = "",
    /**
     * The other addresses this same account signs in at: `login.microsoftonline.com` for an
     * `office.com` entry, a company's SSO, another country's extension. The anti-phishing check
     * accepts them exactly as it accepts [url] — see `DomainMatch.check`.
     */
    val otherDomains: List<String> = emptyList(),
    val totpSecret: String = "",
    val notes: String = "",
    val isFavorite: Boolean = false,
    val cardholderName: String = "",
    val cardNumber: String = "",
    val cardExpiry: String = "",
    val cardCvv: String = "",
    val cardPin: String = "",
    val cardIssuer: String = "",
    val createdAt: DartDateTime,
    val updatedAt: DartDateTime,
) {
    /** Never the fields: a data class would print the password into any log or crash report. */
    override fun toString(): String = "Entry(id=$id, type=$type)"

    companion object {
        const val DEFAULT_CATEGORY = "Autres"

        /**
         * How many [otherDomains] an entry may carry. The anti-phishing check measures an edit
         * distance against each of them before every password copy, so an unbounded list — pasted
         * by hand, or imported from a vault that holds hundreds of addresses — would be paid for on
         * the one action that must stay instant.
         */
        const val MAX_OTHER_DOMAINS = 16
    }
}
